### ISSUE TITLE
Code cleanup of src/sas/sascalc/pr/p_invertor.py

### ISSUE THREAD

**Issue by llimeht** [SasView/sasview#2417]
**Describe the bug**
`src/sas/sascalc/pr/p_invertor.py` contains some code that is not particularly pythonic. Some of this may be a clever and subtle design (that I'd like us to document better so the next person doesn't scratch their head for so long), some of it may be just a bit odd, and some of it is buggy.

This issue report is to try to enumerate the things that stood out while looking at numpy 1.24 compatibility (see #2416) and to figure out what should change.

 1. the `Pinvertor` class starts off using class attributes for `x`, `y`, `err`, `npoints`, `ny`, `nerr`, `slit_height`, `slit_width` but then converts them to instance attributes in later functions (see terminology footnote).  This might be a very clever and subtle way of reducing some memory usage if multiple instances are needed, a hangover from a different approach, or unintended. Whichever it is, the reader needs to think very carefully about how future assignments are done and whether they are mutating the class attribute or setting a new instance attribute that shadows the class attribute; that seems unnecessary. 
 2. the type declared for each of `x`, `y` etc is different to what is used later — later code insists that these are `np.float64` but the initial code only sets them to be `dtype=np.float` which may not be the same depending on the machine. 
 3. many of the setter functions (`set_x`, `set_y` etc) poke around with `__dict__` in unusual and unnecessary ways
 4. many of the getter functions (`get_x`, etc) have the line `data = np.float64(data)` in them. This is ostensibly to do some type marshalling (these functions are old-style functions where the target for the output of the function is passed as an argument to the function). If the type of `data` is correct, then this is a no-op and the `data` variable points at the same numpy array; if the type is not correct (say `dtype=np.float32`) then a new numpy array is created and the data is not actually exported as promised. Either way, this line is bad.

**Suggested fixes**
1. Move the variable definitions from the class block into `__init__`.
2. Change the dtype to `np.float64` uniformly (this is needed anyway, see #2416)
3. Change the poking around with `__dict__` to simple class variable assignments
4. Remove the `data = np.float64(data)` lines.

There's a further question about changing the getters/setters to instead use `@property` decorators; this is a bigger change to the API and lower priority than understanding the above items. Changing the API would also make it possible to avoid copying the data in each of the `get_x`, `get_y` functions which is a potential speed up, but I'm not considering that now.

**What next?**
 * Are all of the above correct?
 * Is @zattala still around to help us understand this code? (I ask this based on git history - the code itself might be older and have other authors, of course!)
 * Do we have a feeling for how good the test suite is for these functions?
 * Is there some benchmarking code already in existence to test that changes don't accidentally break numba jit or cause a performance regression?
 * Does the test suite test that njit is working?

Terminology footnote: For those who haven't played with this particular corner of python land... Class attributes are variables that are shared between all instances of the class when it has been instantiated more than once; they are effectively a global variable within that class. Instance attributes are only for the individual instance and are not shared. Class attributes are rarely what is wanted; they make sense for global-like parameters such as a help URL or copyright statement, while for a calculator object that is operating on different data, the instance attribute is the right thing. Converting between class and instance attributes is messy, adds places where subtle bugs can hide, and offers no advantages over simply defining all instance attributes in `__init__`.

**Comment by lucas-wilkins** [comment]
> Move the variable definitions from the class block into __init__.

Yes, that looks like a reasonable thing to do.

I'd also suggest changing set and get methods to property getters and setters.

> Change the dtype to np.float64 uniformly (this is needed anyway, see https://github.com/SasView/sasview/issues/2416)

64 bit seems like the go-to size these days (but see below)

> Change the poking around with __dict__ to simple class variable assignments

Yes

> Remove the data = np.float64(data) lines.

I think having calculations done with the same bit depth is a good idea in general. Probably unnecessary in most cases, but I would suggest its actually best practice have this kind of consistency in scientific applications. Why do you want to get rid of them?

**Comment by llimeht** [comment]
> I'd also suggest changing set and get methods to property getters and setters.

Perhaps. It would certainly be more pythonic to do so. That's a significant amount of work and an API break that I'm unlikely to do, particularly without knowledge of how good the test suite is.

> > Remove the data = np.float64(data) lines.

> I think having calculations done with the same bit depth is a good idea in general. Probably unnecessary in most cases, but I would suggest its actually best practice have this kind of consistency in scientific applications. Why do you want to get rid of them?

The calculations should certainly be done with the correct precision but this approach doesn't work. 

Let's start by unpacking that single line:

```python
data = np.float64(data)
```
 - if the precision of `data` is `np.float64` then this is a no-op, the Python name `data` continues to point at the same numpy array object as it did before
 - else (when the precision of `data` is not `np.float64`) a new numpy array is created and the `data` name now points to this newly created object.

Now let's look at what that means in the context of the function:

```python
    def get_x(self, data):
        """
        Function to get the x data.

        :param data: Array to place x into
        :return: npoints - Number of entries found.
        """
        data = np.float64(data)
        ndata = len(data)

        if ndata < self.npoints:
            logger.error("Pinvertor.get_x: input array too short for data.")
            return None

        data[:] = self.x

        return self.npoints

```

The important thing is that the function is passing its calculation _output_ back via the `data` that was _passed_ _into_ the `get_foo` function; the `get_foo` functions do not `return` the calculation outputs. Consequently, if you change `data` to point to a different numpy array (which is what `data = np.float64(data)` does if the precision is incorrect!), then all later operations on `data` are wasted as they do not apply to the `data` provided by the caller..

If the calling code has made sure the precision is correct, then `data = np.float64(data)` is just a few wasted CPU cycles; if the calling code gets the precision wrong, then `data = np.float64(data)` prevents the function from returning the requested calculation.

**Comment by lucas-wilkins** [comment]
OK, yeah, I should have looked at the whole function, not just the start.

Something does need to be done. It's quite a bit of work to do right, and that removing the calls to float64 would be the simplest fix for the worst problems. Then again, it seems like nobody has ever called it with anything other that a float64, or they would have noticed.

I don't think there's much efficiency to be gained in trying to wite C in python like is done here, and the get and set methods are a bit weird anyway. I could have a go at re-writing it.

**Comment by lucas-wilkins** [comment]
Anything that changes the interface is currently blocked by #2181 (I think)
